packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -459,7 +534,7 @@ class TextPainter {
   // TODO(garyq): Use actual extended grapheme cluster length instead of
   // an increasing cluster length amount to achieve deterministic performance.
   Rect _getRectFromUpstream(int offset, Rect caretPrototype) {
```

Comment:
This method and the next terrify me. We really should find a way to solve this problem without searches and flattening strings and so on.